classes/observation_manager.php

```diff
@@ -0,0 +1,402 @@
+        }
+        return [$observation, $course, $cm];
+    }
+
+    /**
+     * Gets observation, course and coursemodule from observation instance ID
+     * @param int $obid Observation instance ID
+     * @param string $tablename Database table name
+     * @return list List containing the observation instance, course and coursemodule (in that order)
+     */
+    public static function get_observation_course_cm_from_obid(int $obid, string $tablename = 'observation') {
+        global $DB;
+        if (!$cm = get_coursemodule_from_instance($tablename, $obid)) {
+            throw new \moodle_exception('invalidcoursemodule');
+        }
+        list($course, $cm) = get_course_and_cm_from_cmid($cm->id, 'observation');
+        if (!$observation = $DB->get_record($tablename, ['id' => $obid])) {
+            throw new \moodle_exception('moduleinstancedoesnotexist');
+        }
+        return [$observation, $course, $cm];
+    }
+
+    /**
+     * Modifies an instance of an observation by either creating a new one or updating existing one.
+     * Note that when updating an instance, an ID must be passed in the $data param array.
+     * @param mixed $data Data to be passed to the update or create DB function
+     * @param bool $newinstance If true creates new instance, else updates instance.
+     * @param string $tablename Name of database table to operate on.
+     * @return int If param $newinstance is true, returns ID of new instance. Else returns 1 if updated successfully, else 0.
+     */
+    public static function modify_instance($data, bool $newinstance = false, string $tablename = 'observation'): int {
+        global $DB;
+
+        // Editor data need to be checked to ensure empty strings are not added.
+        if ($data['observer_ins'] === "") {
+            $data['observer_ins'] = null;
+            $data['observer_ins_f'] = null;
+        }
+
+        if ($data['observee_ins'] === "") {
+            $data['observee_ins'] = null;
+            $data['observee_ins_f'] = null;
+        }
+
+        if ($newinstance) {
+            return $DB->insert_record($tablename, $data);
+        } else {
+            return (int)$DB->update_record($tablename, $data);
+        }
+    }
+
+    /**
+     * Modifies or creates a new observation point in the database
+     * @param mixed $data Data to pass to database function
+     * @param bool $newinstance True if new instance, else false if editing
+     * @param bool $returnid True if should return id (only when $newinstance = true)
+     * @param string $tablename Tablename
+     * @return mixed True if successful. If $returnid is True and $newinstance is True, returns ID
+     */
+    public static function modify_observation_point($data, bool $newinstance = false, bool $returnid = false,
+            string $tablename = 'observation_points') {
+        global $DB;
+
+        $data = (object)$data;
+
+        if (property_exists($data, 'max_grade')) {
```

Comment:
With property_exists, this could potentially be widened to if !empty() which checks for null, empty string and whether it bool coerces to false, which might be a little more robust